src/array.c

```diff
@@ -1,24 +1,286 @@
+}
+
+// Set the value at the specified index
+void set(DynamicArray *dy_arr_ptr, int index, void *value) {
+    checkPointerAndIndex(dy_arr_ptr, index);
+
+    size_t element_size = getDataTypeSize(dy_arr_ptr->type);
+    void *destination = (char *)dy_arr_ptr->data + index * element_size;
+    memcpy(destination, value, element_size);
+}
+
+// Get the current size of the dynamic array
+size_t size(DynamicArray *dy_arr_ptr) {
     if (dy_arr_ptr == NULL) {
-        fprintf(stderr, "Failed to allocate memory for dynamic array.\n");
+        fprintf(stderr, "Dynamic array pointer is NULL.\n");
         exit(EXIT_FAILURE);
-    };
+    }
+
+    return dy_arr_ptr->size;
+}
+
+// Clear the dynamic array (set size to 0)
+void clear(DynamicArray *dy_arr_ptr) {
+    if (dy_arr_ptr == NULL) {
+        fprintf(stderr, "Dynamic array pointer is NULL.\n");
+        exit(EXIT_FAILURE);
+    }
+
+    // Reset the size to 0
+    dy_arr_ptr->size = 0;
+
+    // Resize the data array to the initial capacity to reclaim memory
+    resize(dy_arr_ptr, INITIAL_CAPACITY);
+}
+
+// Initialize a dynamic array with the given initial capacity and data type
+DynamicArray *initialize(size_t initial_capacity, DataType type) {
+    // Check if the specified data type is valid
+    switch (type) {
+        case INT:
+        case FLOAT:
+        case DOUBLE:
+        case CHAR:
+            break; // Valid data type
+        default:
+            handleError("Error: Invalid data type.");
+            return NULL;
+    }
 
-    dy_arr_ptr->data = (int *)malloc(initial_capacity * sizeof(int));
+    DynamicArray *dy_arr_ptr = malloc(sizeof(DynamicArray));
+    if (dy_arr_ptr == NULL) {
+        handleError("Error: Failed to allocate memory for dynamic array.");
+        return NULL;
+    }
+
+    dy_arr_ptr->data = calloc(initial_capacity, getDataTypeSize(type));
     if (dy_arr_ptr->data == NULL) {
         free(dy_arr_ptr);
-        fprintf(stderr,
-                "Failed to allocation memory for dynamic array data.\n");
-        exit(EXIT_FAILURE);
-    };
+        perror("Error: Failed to allocate memory for dynamic array data.");
+        return NULL;
+    }
 
     dy_arr_ptr->size = 0;
     dy_arr_ptr->capacity = initial_capacity;
+    dy_arr_ptr->type = type;
     return dy_arr_ptr;
-};
+}
+
+// Resize the dynamic array to the new capacity
+bool resize(DynamicArray *dy_arr_ptr, size_t new_capacity) {
+    void *new_data = realloc(dy_arr_ptr->data, new_capacity * getDataTypeSize(dy_arr_ptr->type));
+    if (new_data == NULL) {
+        fprintf(stderr, "Failed to resize dynamic array.\n");
+        return false;
+    }
+    dy_arr_ptr->data = new_data;
+    dy_arr_ptr->capacity = new_capacity;
+    return true;
+}
+
+// Destroy the dynamic array and free memory
+void destroy(DynamicArray *dy_arr_ptr) {
+    if (dy_arr_ptr != NULL) {
+        free(dy_arr_ptr->data);
+        free(dy_arr_ptr);
+        dy_arr_ptr = NULL; // Set pointer to NULL after freeing memory
```

Comment:
this affect only local copy of dy_arr_ptr, this assing not needed ^_^

Reply:
I was simply spending an hour expanding on what primebrook had done on stream and left this here as more of a reference than anything funcitonal par se.